### PR TITLE
Enable cache time-based expiry by default

### LINKED ISSUE #11787: Enable cache time-based expiry by default
The 'new' time-based expiry for caches in Synapse seems to have helped out with memory usage (empirically, based on what people are telling me).
It seems like it would be worth enabling by default so that people don't have to know about this obscure option to get the benefits.

Related: #11641 (it's not obvious that even though Synapse's memory usage only grows by default, it is capable of trimming back down).

### PR DESCRIPTION
**Pull request by H-Shay**
Changes the default config option for time based cache expiry from None to 30m. Fixes #11787, where you will also find more discussion on this change.

### CODE REVIEW

File: changelog.d/11849.misc
Comment (anoadragon453) [inline changelog.d/11849.misc:1]: It's not clear to me here that my existing Synapse installation won't start expiring caches based on time. Perhaps something like:
```suggestion
Enable cache time-based expiry by default for newly generated homeserver config files.
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Note that this only enables cache expiry by default for newly generated config files, not existing installations. This is good because:

1. We don't want to change settings out from under people.
2. There isn't a way to differentiate between "someone's never set this option" and "someone explicitly disabled this option".

So this looks like the right solution to me.

**Comment by reivilibre** [comment]
> Note that this only enables cache expiry by default for newly generated config files, not existing installations. This is good because…

I actually am not so sure (and it's not the impression I got from the meeting yesterday where we mentioned this).

I think we should enable it by default for existing configs because the original point about it being an obscure option still stands and generally homeservers should benefit from this change.

> There isn't a way to differentiate between "someone's never set this option" and "someone explicitly disabled this option".

Setting it to null would explicitly disable the option; on the other hand, once we ship a default config file with options that aren't the default in the config loaders, we lose the ability to distinguish '30 minutes, explicit' and 'no preference'.
We also might cause strange behaviour later on if we decide to remove this option and start showing deprecation warnings and errors to operators *who never set the option manually anyway*.

I find it strange that we would enable an option by default only for 'new' homeservers — I appreciate the caution but it also conflicts with the intuition that a homeserver admin only needs to have the options they changed present in their config file.

By leaving an option unmarked in the config, I would interpret this to mean 'I, the admin, don't care and want the software to do the default reasonable behaviour' — in this case I would say the time-based expiry is a great boon for memory usage and seems like the most reasonable behaviour when the admin hasn't expressed a preference.

There is a case to be made that we should stop documenting the options inside the config file and instead have a nice config manual or something, because the comment will become out of date for old installations. Still, we could announce that in the upgrade notes in case anyone was caught out by the change in default behaviour.

**Comment by H-Shay** [comment]
Fascinating-it never occurred to me that we would change the behavior of existing Synapse homeservers. My understanding was something along the lines of: "we've used this setting and feel confident it's beneficial, we will make it the default going forwards." My personal feeling is that it is a little strange to change the default behavior out from under people rather than having them explicitly op-in to the *new* default (which I figured we would shout about in the release notes) but that's just a feeling. I am happy to implement whatever the consensus determines is the best approach.  
Also plus one for having a config manual vs comments, I will open an issue about that as I think it would be beneficial.

**Comment by richvdh** [comment]
> My personal feeling is that it is a little strange to change the default behavior out from under people rather than having them explicitly op-in to the new default (which I figured we would shout about in the release notes) but that's just a feeling. 

This is certainly a reasonable argument, but I think the point is that most people don't want to have to think this hard about maintaining their deployments - they want it to "just work" and it's up to us to make it do so as best we can, which includes changing the default behaviour where we have reason to believe that will improve things for the majority of users.

There could be a *minority* who would prefer to retain the existing behaviour, but they can update their config files if they want to.

It's worth noting that we change Synapse's default behaviour all the time when it comes to bugfixes and the like. You could argue that not having time-based expiry enabled is a bug.

**Comment by anoadragon453** [comment]
Thanks for the course-correction folks! There may be a small user-base that read the option and left it commented, assuming that would remain disabled. We would then be changing the behaviour for them.

However, shouting about it in the changelog (or perhaps upgrade notes if we deem necessary) should be enough for even this theoretical edge case.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Going forward on this PR, I think we should:

1. Change the default (when the option is not specified) of `expiry_time` to `30m`.
2. Assuming we'd like to continue allowing people to disable cache expiry, modify the sample config to make disabling the option a little simpler. We could either:
    * tell people to write `expiry_time: null` to disable the option, but this isn't really consistent with the rest of the config file.
    * introduce a new `expire_caches` boolean option (defaulting to `true`), which would be my personal preference.

**Comment by callahad** [comment]
> tell people to write `expiry_time: null` to disable the option, but this isn't really consistent with the rest of the config file.

Would `expiry_time: 0` or `expiry_time: -1` be acceptable? Those seem to be common enough conventions in other software, but I don't see much precedent within Synapse. Adding a whole 'nother option seems pretty heavyweight, so I'd bias toward overloading the time.

**Comment by richvdh** [comment]
`expiry_time` is a pretty rubbish name for it, tbh (cf https://github.com/matrix-org/synapse/issues/5584#issuecomment-509284013). It might be a reasonable time to sort that out.

**Comment by H-Shay** [comment]
Thanks everyone for chiming in, I will make these changes. Also thanks for taking time to explain your reasoning, it's very helpful for me to hear.

**Comment by H-Shay** [comment]
Per Rich's comment I renamed the user-facing flag to something (hopefully) a little more descriptive, but I wasn't sure if I should rename the internal var as well.

File: synapse/config/cache.py
Comment (anoadragon453) [inline synapse/config/cache.py:226]: Since we're changing config options, we should include backwards-compatibility support for the previous config option name. Thus if someone set `expiry_time` previously, that same value should continue to work even after this update. At some point :tm: in the future we can remove support for the old option name.

File: synapse/config/cache.py
Comment (anoadragon453) [inline synapse/config/cache.py:233]: May be worth adding a quick comment to indicate that this equates to 30 minutes.

File: synapse/config/cache.py
Comment (anoadragon453) [inline synapse/config/cache.py:225]: We could also set the default here, via:

```suggestion
        cache_entry_ttl = cache_config.get("cache_entry_ttl", "30m")
```

which may be slightly more intuitive if one is scanning the code looking for default option values.

File: synapse/config/cache.py
Comment (anoadragon453) [inline synapse/config/cache.py:153]:
```suggestion
          # Controls whether cache entries are evicted after a specified time
          # period. Defaults to true.
          #
          # expire_caches: false
```

Comment (anoadragon453) [reply synapse/config/cache.py:153]: This block should also have one of those "Uncomment to disable this feature" sentences.

File: changelog.d/11849.misc
Comment (anoadragon453) [inline changelog.d/11849.misc:1]: It may be worth mentioning the new configuration options that are available, or otherwise pointing to the relevant section of the sample config for the full details.

This could also be done via our [upgrading docs](https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md), where we have more room.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
> I wasn't sure if I should rename the internal var as well.

Doing so is fine, as long as it doesn't affect anyone relying on Synapse (such as methods exposed by the module API. These variables are private however, and people shouldn't be reaching into and relying on our private variables.

For Synapse devs, if an in-flight PR relied on the old variable name then they'd discover a merge conflict after your PR had merged - alerting them, and allowing them to correct it.

A few things below, but overall this is looking great!

File: docs/upgrade.md
Comment (reivilibre) [inline docs/upgrade.md:95]: I don't know why there's this extra section above `v1.53.0` — I would probably bring them down into 1.53 since we haven't released an RC for it yet.

File: changelog.d/11849.misc
Comment (reivilibre) [reply changelog.d/11849.misc:1]: I would likely still sneak a keyword or two in there: `expiry_time` will be superseded by `expire_caches` and `cache_entry_ttl`.
Helps perk people's ears up and helps if you ever need to do a Ctrl+F for one of the config option names.

File: synapse/config/cache.py
Comment (reivilibre) [inline synapse/config/cache.py:249]: This is very nitty, but I would call these *options* rather than *flags*.
At least in my mind, a flag is a boolean.

File: docs/upgrade.md
Comment (reivilibre) [inline docs/upgrade.md:90]: I know it's not your mess, but as you're here, please could you clean this up by removing the 'v1.(next)' heading and bringing the announcement into the v1.53 section?
I personally think a 'v1.(next)' section would be easily forgotten as it's not in our release notes. Or at least, I'd forget! :)

**Review by reivilibre — APPROVED** [review]
Looks good; thanks!

File: docs/upgrade.md
Comment (H-Shay) [reply docs/upgrade.md]: sure thing

File: synapse/config/cache.py
Comment (H-Shay) [reply synapse/config/cache.py]: This is a useful distinction, thank you!

**Review by reivilibre — APPROVED** [review]

**Comment by H-Shay** [comment]
@anoadragon453 any chance I can get you to dismiss your stale review when you get the chance?

**Comment by anoadragon453** [comment]
> @anoadragon453 any chance I can get you to dismiss your stale review when you get the chance?

Of course, and note that it's fine to do so yourself as well if the review is outdated (assuming you have permission?).

File: changelog.d/11849.misc
Comment (richvdh) [inline changelog.d/11849.misc:1]: this should probably be a `.feature`?

and "will be" should be "has been"?

File: docs/upgrade.md
Comment (richvdh) [inline docs/upgrade.md:119]: Might be worth mentioning that `expiry_time` will still work (for now), but is deprecated?

**Review by richvdh — COMMENTED** [review]
sorry, a few belated notes. Feel free to commit fixes straight to develop.

This seems to add an empty https://github.com/matrix-org/synapse/blob/develop/synapse/config/background_updates.py, which looks wrong. Please could it be removed?

**Comment by H-Shay** [comment]
Thanks for catching these mistakes-I have made the requested changes.

**Comment by clokep** [comment]
> Thanks for catching these mistakes-I have made the requested changes.

For cross-referencing: 9c4563c5cd97dbcfe3eb43e05b30ca3c830d6937

**Pull request merged**
